File: server/catches_test.go
package main

import "testing"

func catchFixture() State {
	s := initial()
	for i := range s.Players {
		s.Players[i].X = 10
		s.Players[i].Z = 8
	}
	s.Ball = Ball{Owner: -1, LastTouch: 7, H: .25}
	s.Controlled = [2]int{7, 16}
	return s
}
func TestCatchSelectedPlayerRangeAndOrder(t *testing.T) {
	for _, distance := range []int{16, 17} {
		s := catchFixture()
		s.Players[7].X = float64(distance) * terrainUnit
		s.Players[7].Z = 0
		s.catchBall()
		expected := -1
		if distance == 16 {
			expected = 7
		}
		if s.Ball.Owner != expected {
			t.Fatal("catch range", distance)
		}
	}
	s := catchFixture()
	s.Players[8].X = 0
	s.Players[8].Z = 0
	s.catchBall()
	if s.Ball.Owner != -1 {
		t.Fatal("unselected catcher")
	}
	s = catchFixture()
	s.Players[7].X = 0
	s.Players[7].Z = 0
	s.Players[16].X = 15 * terrainUnit
	s.Players[16].Z = 0
	s.catchBall()
	if s.Ball.Owner != 16 {
		t.Fatal("interleaved roster priority")
	}
}
func TestThrowerCannotCatchButOpponentCan(t *testing.T) {
	s := catchFixture()
	s.Players[7].X = 0
	s.Players[7].Z = 0
	s.Players[7].Action = 3
	s.Ball.Lock = .18
	s.catchBall()
	if s.Ball.Owner != -1 {
		t.Fatal("throwing player caught")
	}
	s.Players[16].X = 0
	s.Players[16].Z = 0
	s.catchBall()
	if s.Ball.Owner != 16 {
		t.Fatal("opponent blocked by global throw lock")
	}
}

func TestInterceptionCues(t *testing.T) {
	for _, period := range []int{1, 2} {
		for team := 0; team < 2; team++ {
			for _, friendly := range []bool{false, true} {
				for _, charged := range []bool{false, true} {
					s := catchFixture()
					s.Period = period
					i := team*9 + 7
					s.Players[i].X = 0
					s.Players[i].Z = 0
					throwTeam := 1 - team
					if friendly {
						throwTeam = team
					}
					s.Ball.LastTouch = throwTeam*9 + 6
					s.Ball.Charged = charged
					s.Ball.Electric = 0
					s.catchBall()
					if s.Ball.Owner != i {
						t.Fatal("catch owner")
					}
					count := 0
					for _, e := range s.Events[:s.EventCount] {
						if e.Kind == 24 || e.Kind == 25 {
							count++
							if e.Kind != 24+team {
								t.Fatal("team cue")
							}
						}
					}
					expected := 0
					if !friendly && !charged {
						expected = 1
					}
					if count != expected {
						t.Fatal("interception count", team, friendly, charged, count)
					}
				}
			}
		}
	}
}

func TestCatchImpactRequiresPlanarMotion(t *testing.T) {
	for _, velocity := range [][2]float64{{0, 0}, {1, 0}, {0, -1}} {
		for _, enemy := range []bool{false, true} {
			s := catchFixture()
			s.Players[7].X, s.Players[7].Z = 0, 0
			s.Ball.VX, s.Ball.VZ, s.Ball.VH = velocity[0], velocity[1], -1
			s.Ball.LastTouch = 6
			if enemy {
				s.Ball.LastTouch = 16
			}
			s.catchBall()
			if s.Ball.Owner != 7 {
				t.Fatal("catch failed")
			}
			impact, intercept := false, false
			for _, e := range s.Events[:s.EventCount] {
				impact = impact || e.Kind == 16
				intercept = intercept || e.Kind == 24
			}
			if impact != (velocity[0] != 0 || velocity[1] != 0) || intercept != enemy {
				t.Fatal("incorrect cues", velocity, enemy)
			}
		}
	}
}

File: src/catches.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { initial, catchBall } from "./game.js";
const unit = 22.4 / 576;
function setup() {
  const s = initial();
  s.players.forEach((p) => Object.assign(p, { x: 10, z: 8 }));
  Object.assign(s.ball, {
    x: 0,
    z: 0,
    h: 0.25,
    owner: -1,
    lastTouch: 7,
    lock: 0,
  });
  s.controlled = [7, 16];
  return s;
}
test("only selected players can catch, with original 16-unit inclusive range", () => {
  for (const distance of [16, 17]) {
    const s = setup();
    Object.assign(s.players[7], { x: distance * unit, z: 0 });
    catchBall(s);
    assert.equal(s.ball.owner, distance === 16 ? 7 : -1);
  }
  const s = setup();
  Object.assign(s.players[8], { x: 0, z: 0 });
  catchBall(s);
  assert.equal(
    s.ball.owner,
    -1,
    "unselected teammate must not steal the catch",
  );
});
test("throwing player cannot catch, but another selected player can intercept immediately", () => {
  const s = setup();
  Object.assign(s.players[7], { x: 0, z: 0, action: 3 });
  s.ball.lock = 0.18;
  catchBall(s);
  assert.equal(s.ball.owner, -1);
  Object.assign(s.players[16], { x: 0, z: 0 });
  catchBall(s);
  assert.equal(s.ball.owner, 16, "no invented global post-throw catch lock");
});
test("contested catches follow interleaved roster order, team two first at equal index", () => {
  const s = setup();
  Object.assign(s.players[7], { x: 0, z: 0 });
  Object.assign(s.players[16], { x: 15 * unit, z: 0 });
  catchBall(s);
  assert.equal(
    s.ball.owner,
    16,
    "earlier eligible player wins, not the nearest",
  );
  const other = setup();
  other.controlled[0] = 6;
  Object.assign(other.players[6], { x: 15 * unit, z: 0 });
  Object.assign(other.players[16], { x: 0, z: 0 });
  catchBall(other);
  assert.equal(other.ball.owner, 6);
});

test("interception cue identifies the team and excludes friendly or charged catches", () => {
 for(const team of [0,1]) for(const friendly of [false,true]) for(const charged of [false,true]) {
  const s=setup(),i=team*9+7;
  Object.assign(s.players[i],{x:0,z:0});
  Object.assign(s.ball,{lastTouch:(friendly?team:1-team)*9+6,charged,electric:0});
  catchBall(s);
  assert.equal(s.ball.owner,i);
  const intercepts=s.events.filter(e=>e.kind===24||e.kind===25);
  assert.equal(intercepts.length,!friendly&&!charged?1:0);
  if(intercepts.length) assert.equal(intercepts[0].kind,24+team);
 }
});
